### ISSUE TITLE
proposal: Support inlay hint feature

### ISSUE THREAD

**Issue by septemhill** [gleam-lang/gleam#2319]
Hi gleam team,

I would like to suggest support `inlay hint` feature in vscode-gleam. 
This feature be supported in most of popular programming languages (e.g. Rust, Go, etc.) 
It would save us much time to hover over the variables/functions to check the types.

Looking forward to hearing from your reply !

**Comment by inoas** [comment]
I like this IF it is something I can disable per project via vscode settings.

**Comment by septemhill** [comment]
> Seeing as Gleam has labels does it still make sense to show the argument names in function calls?

Agree. For this case, I think we could ignore. But another question is, if the function without named arguments how do we display, may be just show the type ?

**Comment by lpil** [comment]
Fab! Ok so the next steps are to determine how to send these hints via the language server protocol, and how to turn them on and off with the language server protocol.

**Comment by inoas** [comment]
> > Are there any other places we might want them? The language server can put them anywhere.
> 
> This proposal is all I can think of now.
> 
> > Seeing as Gleam has labels does it still make sense to show the argument names in function calls?
> 
> Agree. For this case, I think we could ignore. But another question is, if the function without named arguments how do we display, may be just show the type ?
> 
> I understand we don't want to make whole UI with too much noise. So here I provide Go inlay-hint features in LSP settings. There're many hints in Go and we could decide which hint would be enabled in a language rather than a project. But the config could be created under `.vscode/setting` in different projects.
> 
> ![Screenshot 2023-09-04 at 9 13 21 AM](https://user-images.githubusercontent.com/1216949/265299767-ae3f5324-7ebf-4cb8-a90a-19afeadc2dbf.png)
> 
> Thanks.

would we want the comments around it, i think the rust lsp skips those. just grey and some opacity, and maybe (not sure) you cannot select them ever with the cursor

**Comment by lpil** [comment]
We don't have an inline comment syntax so we couldn't wrap them like Go does.

**Comment by jamesbirtles** [comment]
I'm currently working on this https://github.com/jamesbirtles/gleam/tree/feat/inlay-hints

**Comment by lpil** [comment]
Very exciting!

Let's start with the MVP feature-set and we can add more over time. That way it is easier to review and merge.

**Comment by lpil** [comment]
https://github.com/gleam-lang/gleam/pull/2393

**Comment by GearsDatapacks** [comment]
Also another thing inlay hints could be used for is inferred types for arguments and return-types, since those might not be obvious in certain cases

**Comment by lpil** [comment]
For sure!!

**Comment by codemonkey76** [comment]
> I'm currently working on this https://github.com/jamesbirtles/gleam/tree/feat/inlay-hints along with some code actions to add in the types when clicked
> 
> ![CleanShot 2023-11-02 at 20 39 23@2x](https://private-user-images.githubusercontent.com/3743418/280135250-53d09b04-cfcf-4577-a3ab-f9e2649d734f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjAwNTQ1NjksIm5iZiI6MTcyMDA1NDI2OSwicGF0aCI6Ii8zNzQzNDE4LzI4MDEzNTI1MC01M2QwOWIwNC1jZmNmLTQ1NzctYTNhYi1mOWUyNjQ5ZDczNGYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcwNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MDRUMDA1MTA5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MGEyNmU2ZjFkNzExN2JlZDMwNjk5MDgwYWNmZWFmMDk3NTA1MjlkN2VmMDhkMjlmOTRiMTVhYzg2MWI4NzllOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.is_izw-3xNjGQOHi78dxhGKkgn9aHh0xO-3frxg-l4w)
> 
>  CleanShot.2023-11-02.at.20.59.20.mp4 
>  CleanShot.2023-11-02.at.21.44.31.mp4 
> Currently working for top level functions and constants, its a little trickier for things inside functions like calls and variable assignment as I don't think theres a generic visitor/walker(?) to get at all expressions without manually recursing through them

I get a panic when using your feat/inlay-hints:
/home/shane/dev/rust/gleam/compiler-core/src/language_server/server.rs:565
Unable to parse configuration: invalid type: null, expected struct Configuration

**Comment by JasonBoyett** [comment]
has anyone tried to get inlay hints working with NeoVim or other non vs code editors? If so can you point me in the right direction?
